# Hola 😀,

Soy **Hesus Garcia** – **"Soy el único Hesus que conoces (y probablemente conocerás) 🌟"** – Sí, como "Jesús", pero con una H que me hace único. Puede sonar raro, pero créeme, ¡no lo olvidarás! Como tu revisor en Triple-Ten, estoy aquí para guiarte y ayudarte a mejorar tu código. Si algo necesita un ajuste, no hay de qué preocuparse; ¡aquí estoy para hacer que tu trabajo brille con todo su potencial! ✨
Cada vez que encuentre un detalle importante en tu código, te lo señalaré para que puedas corregirlo y así te prepares para un ambiente de trabajo real, donde el líder de tu equipo actuaría de manera similar. Si en algún momento no logras solucionar el problema, te daré más detalles para ayudarte en nuestra próxima oportunidad de revisión.

Es importante que cuando encuentres un comentario, **no los muevas, no los modifiques, ni los borres**.

---

### Formato de Comentarios

Revisaré cuidadosamente cada implementación en tu notebook para asegurar que cumpla con los requisitos y te daré comentarios de acuerdo al siguiente formato:

<div class="alert alert-block alert-success">
<b>Comentario del Revisor - 1ª Iteración</b> <a class="tocSkip"></a><br>
<b>Éxito</b> - ¡Excelente trabajo! Esta parte está bien implementada y contribuye significativamente al análisis de datos o al proyecto. Continúa aplicando estas buenas prácticas en futuras secciones.
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del Revisor - 1ª Iteración</b> <a class="tocSkip"></a><br>
<b>Atención</b> ⚠️ - Este código está correcto, pero se puede optimizar. Considera implementar mejoras para que sea más eficiente y fácil de leer. Esto fortalecerá la calidad de tu proyecto.
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del Revisor - 1ª Iteración</b> <a class="tocSkip"></a><br>
<b>A resolver</b> ❗ - Aquí hay un problema o error en el código que es necesario corregir para aprobar esta sección. Por favor, revisa y corrige este punto, ya que es fundamental para la validez del análisis y la precisión de los resultados.
</div>

---

Justo después de este mensaje encontrarás el **Comentario General del Revisor**, que incluirá:

- **Aspectos positivos:** Un resumen de los puntos fuertes de tu proyecto.
- **Áreas de mejora:** Sugerencias sobre aspectos donde puedes mejorar.
- **Temas adicionales para investigar:** Ideas de temas opcionales que puedes explorar por tu cuenta para desarrollar aún más tus habilidades.

Estos temas adicionales no son obligatorios en esta etapa, pero pueden serte útiles para profundizar en el futuro.

---

Esta estructura en viñetas facilita la lectura y comprensión de cada parte del comentario general.

También puedes responderme de la siguiente manera si tienes alguna duda o quieres aclarar algo específico:

<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>
Aquí puedes escribir tu respuesta o pregunta sobre el comentario.
</div>

**¡Empecemos!** 🚀

<div class="alert alert-block alert-success">
<h2>Comentario General del Revisor - 1ª Iteración <a class="tocSkip"></a></h2>

Tu proyecto está <b>aprobado</b>. 🎉

<br><br>

<b>Aspectos positivos:</b>
<ul>
<li>Excelente uso de bootstrap: 1000 iteraciones con reemplazo para estimar intervalos de confianza y cuantificar el riesgo de pérdida.</li>
<li>Buen uso de sklearn y pandas para entrenar modelos en paralelo en tres regiones y comparar resultados.</li>
<li>Muy buen reporte de métricas: volumen medio, RMSE, desviación estándar, ganancias, riesgo e intervalos de confianza del 95%.</li>
<li>La estructura general del análisis es clara, reproducible y está alineada con los objetivos del proyecto.</li>
</ul>

<b>Recomendaciónes:</b>
<ul>
<li>La conclusión final podría aclarar mejor por qué se prioriza la Región 1 por menor riesgo de pérdida frente a la Región 0, que presenta mayor ganancia estimada. Esto no impide la aprobación, pero fortalecería la argumentación de negocio.</li>
</ul>

¡Felicidades! Has realizado un muy buen trabajo aplicando modelado predictivo, simulación bootstrap y análisis de riesgo para la toma de decisiones. 🚀😊

</div>

A partir de tres conjuntos de datos que representan características y cantidades de barriles de petroleo producidas por pozo en tres diferentes regiones, se realiza el entrenamiento de un modelo de regresión lineal aplicando la función `lineal_regresion_model_trainer()` para cada dataset.

* Se generan las predicciones y los target
* Se analizan los valores de volumen medio y RCME.
    
Se crea la función `calculate_profit()`:

* Argumentos: 
    * La respuesta de `lineal_regresion_model_trainer()`
    * Presupuesto 
    * Cantidad de pozos 
    * precio por unidad (product convertido a dólares)
* Proceso:
    * Transforma los valores 'predictions' y 'target' del diccionario en un dataframe de pandas.
    * Acomoda por ganancia predicha por el modelo de mayor a menor
    * Toma los primeros `total_wells` elementos del Dataframe y calcula el beneficio neto (real) en función del presupuesto.
* Respuesta
    * Ganancia total de un grupo de n pozos `profits`
        
Se utiliza la téctica de `bootstrap` para generar 1000 grupos aleatorios de 500 muestras con reemplazo y diferente `random_state` de las respuestas de la función `lineal_regresion_model_trainer()` eligiendo las 200  mejores según predicciones del modelo y calculando la ganancia real de las mismas con `calculate_profit()`:
* De cada grupo de 500, se obtienen el profit real; En total, mil valores de profit para calcular:
    * Valores promedio de ganancia por grupo de pozos por región.
    * Conteos de ganancias y pérdidas reales del top de valores seleccionados por el modelo
    * Riesgo de Pérdida
    * Intervalo de confianza del 95% para el riesgo de pérdida
* Este proceso completo (bootstrap se repite una vez por cada dataset (archivo csv))
Finalmente se concluye cuál es la mejor región para invertir y por qué.
*Las funciones se encuentran dentro de la carpeta creada 'src' ubicada en el mismo nivel jerárquico que el proyecto*

# Importación de librerías

In [ ]:
import warnings
warnings.filterwarnings("ignore")
import sys
import os
# Le dice python que busque liberrías ahí también
sys.path.append(os.path.join('../src'))

import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
from sklearn.utils import shuffle
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import mean_squared_error
import funciones_personales as fp


# importación de datos

Los datos son sintéticos: los detalles del contrato y las características del pozo no se publican.

**Descripción de datos**

Los datos de exploración geológica de las tres regiones se almacenan en archivos:

* `/datasets/geo_data_0.csv`
* `/datasets/geo_data_1.csv`
* `/datasets/geo_data_2.csv`

Cada uno cuenta con las columnas:

* `id` — identificador único de pozo de petróleo
* `f0`, `f1`, `f2` — tres características de los puntos (su significado específico no es importante, pero las características en sí son significativas)
* `product` — volumen de reservas en el pozo de petróleo (miles de barriles).


In [2]:
df0 = pd.read_csv('../datasets/geo_data_0.csv')
df1 = pd.read_csv('../datasets/geo_data_1.csv')
df2 = pd.read_csv('../datasets/geo_data_2.csv')

In [3]:
# crea la lista de df
lista_dfs= [df0, df1, df2]
for df in lista_dfs:
    print(df.info())
    print(df.head(5))
    print(df.nunique())
    print(df.duplicated().sum())
    print("-"*50)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None
      id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647
id          99990
f0         100000
f1         100000
f2         100000
product    100000
dtype: int64
0
--------------------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 

## Conclusion análisis de datos

No se encontraron valores nulos ni filas complletamente repetidas, se decide continuar con entrenamiento normal y posteriormente con bootstrapping

# Predicciones y targets de las tres regiones
* Cada csv es tranformado en un diccionario que contiene
    * predicciones,
    * target,
    * volumen medio de reservas y
    * RCME del modelo  

In [4]:
# Creación de respuestas y agregados de datos a cada df
df0_resp=fp.linear_regresion_model_trainer(
    dataset=df0, target_col='product',
    feature_cols=['f0', 'f1', 'f2'], verbose=True, random_state=54321)
df1_resp=fp.linear_regresion_model_trainer(
    dataset=df1, target_col='product',
    feature_cols=['f0', 'f1', 'f2'], verbose=True, random_state=54321)
df2_resp=fp.linear_regresion_model_trainer(
    dataset=df2, target_col='product',
    feature_cols=['f0', 'f1', 'f2'], verbose=True, random_state=54321)

Volumen medio de reservas: 92.50
RECM del modelo de regresión lineal en el conjunto de validación: 37.68
El promedio de la desviación estandar es 20.33%
Volumen medio de reservas: 68.83
RECM del modelo de regresión lineal en el conjunto de validación: 0.89
El promedio de la desviación estandar es 0.65%
Volumen medio de reservas: 95.00
RECM del modelo de regresión lineal en el conjunto de validación: 40.15
El promedio de la desviación estandar es 21.13%


## Conclusiones del entrenamiento
* Por razones aún desconocidas el modelo se ajusta especialmente bien para la región 1
* el modelo parece comportarse de forma similar para la región 0 y región 2.

# Cálculo de ganancia
la función recibe el diccionario resultado de la función `linear_regresion_model_trainer()` y calcula la ganancia real, además calcula la variable `min_well_income` que es el número mínimo de unidades (miles de barriles) para que la inversión comience a ser redituable.
* Se definen las variables a usarse para este caso particular

In [5]:
total_budge= 10**8
total_wells=200
unit_income=4500
gan0= fp.calculate_profit(df0_resp, total_budge, unit_income, total_wells)
print(f"Ganancia total (millones de dólares) para el modelo 0: ${gan0/10**6:.2f}")
gan1= fp.calculate_profit(df1_resp, total_budge, unit_income, total_wells)
print(f"Ganancia total (millones de dólares) para el modelo 1: ${gan1/10**6:.2f}")
gan2= fp.calculate_profit(df2_resp, total_budge, unit_income, total_wells)
print(f"Ganancia total (millones de dólares) para el modelo 2: ${gan2/10**6:.2f}")

Ganancia total (millones de dólares) para el modelo 0: $31.79
Ganancia total (millones de dólares) para el modelo 1: $24.15
Ganancia total (millones de dólares) para el modelo 2: $24.14


## Conclusiones y recomendación de región
* Se encontró que el número mínimo por pozo para generar ganancia es 111.1. Ninguna región alcanzó esta cifra para sus valores promedio
    * **Región 0**: Ganancia: $31.79M Volumen medio: 92.50
    * **Región 1**: Ganancia: $24.15M Volumen medio:: 68.83
    * **Región 2**: Ganancia: $24.14M Volumen medio : 95.00

 **Se propone Región 0 como mejor candidato**
 * Mayor ganancia de los tres, segundo mejor volumen medio.
 * Segundo candidato Región 2: mayor volumen de los tres, diferencia 0.01M detrás del segundo mejor.

# Bootstraping

1000 iteraciones con reemplazo para estimar intervalos de confianza y cuantificar riesgo de pérdida.

In [6]:
lista = [df0_resp, df1_resp, df2_resp]
for df, num in zip(lista, range(len(lista))):
    profits=[]
    df = pd.DataFrame(df)[['predictions', 'target']]
    for i in range(1000):
        sample = df.sample(n=500, replace=True, random_state=54321+i)
        best_sample = sample.sort_values('predictions', ascending=False).head(total_wells)
        profit = fp.calculate_profit(best_sample, total_budge, unit_income, total_wells)
        profits.append(profit)
        profit_mean = best_sample['target'].mean()
        #print(f"Ganancia total (millones de dólares) para la muestra {i+1}: ${profit/10**6:.2f}")
    profits=pd.Series(profits)
    profits_positives = profits[profits > 0].count()
    profits_negatives = profits[profits < 0].count()
    risk= profits_negatives / len(profits)
    profit_lower = profits.quantile(.025)
    profit_upper = profits.quantile(.975)
    print(f"Región {num}")
    print(f"Ganancias positivas: {profits_positives}")
    print(f"Ganancias negativas: {profits_negatives}")
    print(f"Riesgo de pérdidas: {risk*100:.2f}%")
    print(f"Intervalo de confianza del 95% para la ganancia: ${profit_lower/10**6:.2f} - ${profit_upper/10**6:.2f}")


Región 0
Ganancias positivas: 926
Ganancias negativas: 74
Riesgo de pérdidas: 7.40%
Intervalo de confianza del 95% para la ganancia: $-1.60 - $9.18
Región 1
Ganancias positivas: 972
Ganancias negativas: 28
Riesgo de pérdidas: 2.80%
Intervalo de confianza del 95% para la ganancia: $-0.07 - $7.86
Región 2
Ganancias positivas: 920
Ganancias negativas: 80
Riesgo de pérdidas: 8.00%
Intervalo de confianza del 95% para la ganancia: $-1.16 - $8.53


# Conclusiones generales:
* En un inicio, los resultados del modelo para cada región mostraron que la región más afín al modelo era la 1:
    * Región 0:
        * Volumen medio de reservas: 92.50
        * RECM en el conjunto de validación: 37.68
        * Promedio de la desviación estandar: 20.33%
    * Región 1:
        * Volumen medio de reservas: 68.83
        * RECM en el conjunto de validación: 0.89
        * Promedio de la desviación estandar: 0.65%
    * Región 2:
        * Volumen medio de reservas: 95.00
        * RECM en el conjunto de validación: 40.15
        * Promedio de la desviación estandar: 21.13%

* Tras el primer análisis de ganancias se deicidió elegir la Región 0:

    * Ganancia (millones de dólares) Región 0: $31.79
    * Ganancia (millones de dólares) Región 1: $24.15
    * Ganancia (millones de dólares) Región 2: $24.14

* Después del bootstrap y el cálculo de confianza y riesgo se decide elegir la región 1 (la misma que al inicio) para invertir:

    * Región 0:
        * Riesgo de pérdidas: 7.40%
        * Intervalo de confianza del 95% para la ganancia: -1.60 , 9.18
    * Región 1:
        * Riesgo de pérdidas: 2.80%
        * Intervalo de confianza del 95% para la ganancia: -0.07 , 7.86
    * Región 2:
        * Riesgo de pérdidas: 8.00%
        * Intervalo de confianza del 95% para la ganancia: -1.16 , 8.53

**Se recomienda invertir en la región 1 debido a que el riesgo de pérdidas, aunque no es menor a 2.5% se encuentra bastante cerca además de que aún en un escenario pesimista la pérdída máxima estimada será de $70,000 dólares del total (~ 0.07% del presupuesto).**

**Se prioriza la Región 1 por menor riesgo de pérdida frente a la Región 0, que presenta mayor ganancia estimada.**